1.18.0 and nightly ICE:  /checkout/src/librustc_trans/mir/constant.rs:377: _1 not initialized

```bash
$ rustc --version
rustc 1.20.0-nightly (bf0a9e0b4 2017-07-10)
$ 
$ rustc tmp.rs
error[E0080]: constant evaluation error
 --> tmp.rs:3:18
  |
3 |    const B:i32 = A[0];	 
  |                  ^^^^ index out of bounds: the len is 0 but the index is 0

error: internal compiler error: /checkout/src/librustc_trans/mir/constant.rs:378: _1 not initialized
 --> tmp.rs:4:12
  |
4 |    let _ = &B;
  |            ^^

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.20.0-nightly (bf0a9e0b4 2017-07-10) running on x86_64-unknown-linux-gnu

thread 'rustc' panicked at 'Box<Any>', /checkout/src/librustc_errors/lib.rs:437:8
note: Run with `RUST_BACKTRACE=1` for a backtrace.

$ 
$ rustc-1.18.0 tmp.rs
error[E0080]: constant evaluation error
 --> tmp.rs:3:18
  |
3 |    const B:i32 = A[0];	 
  |                  ^^^^ index out of bounds: the len is 0 but the index is 0

error: internal compiler error: /checkout/src/librustc_trans/mir/constant.rs:377: _1 not initialized
 --> tmp.rs:4:12
  |
4 |    let _ = &B;
  |            ^^

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

thread 'rustc' panicked at 'Box<Any>', /checkout/src/librustc_errors/lib.rs:375
note: Run with `RUST_BACKTRACE=1` for a backtrace.

$ 
$ cat tmp.rs
fn main () {
   const A:&[i32] = &[];	
   const B:i32 = A[0];	 
   let _ = &B;
} 
$ 
```

Another way to reproduce:
```Rust
fn main() {
    const X: u32 = 0-1;
    println!("{}", X);
}
```
```
warning: constant evaluation error: attempt to subtract with overflow. This will become a HARD ERROR in the future
 --> src/main.rs:3:20
  |
3 |     const X: u32 = 0-1;
  |                    ^^^
  |
  = note: #[warn(const_err)] on by default

error[E0080]: constant evaluation error
 --> src/main.rs:3:20
  |
3 |     const X: u32 = 0-1;
  |                    ^^^ attempt to subtract with overflow

error: internal compiler error: /checkout/src/librustc_trans/mir/constant.rs:378: _1 not initialized
 --> src/main.rs:4:20
  |
4 |     println!("{}", X);
  |                    ^

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.20.0-nightly (f85579d4a 2017-07-12) running on x86_64-unknown-linux-gnu
```

Can replace `0-1` with `33-44` for example, the error is the same. (that `_1` is not related to `1`)
Maybe this helps a bit:
I put the following line in [this rustc code](https://github.com/rust-lang/rust/blob/148718b4f32b46a96e339be92102697ad58b170b/src/librustc_trans/mir/constant.rs#L727-L728):
```patch
diff --git a/src/librustc_trans/mir/constant.rs b/src/librustc_trans/mir/constan
t.rs
index 98e774a298..e80c13cc9b 100644
--- a/src/librustc_trans/mir/constant.rs
+++ b/src/librustc_trans/mir/constant.rs
@@ -725,6 +725,7 @@ impl<'a, 'tcx> MirConstContext<'a, 'tcx> {
             }
 
             mir::Rvalue::Ref(_, bk, ref lvalue) => {
+                eprintln!("!!!!! lvalue={:?} bk={:?} span={:?}",lvalue, bk, spa
n);
                 let tr_lvalue = self.const_lvalue(lvalue, span)?;
 
                 let ty = tr_lvalue.ty;
```
It shows:
`!!!!! lvalue=_1 bk=Shared span=src/main.rs:4:20: 4:21`

<details>
<summary>partial console output/stacktrace</summary>

```rust
    Running2 `rustc --crate-name bug43197 src/main.rs --crate-type bin --emit=dep-info,link -C debuginfo=2 -C metadata=fd80fe04a98d7b0a -C extra-filename=-fd80fe04a98d7b0a --out-dir /home/xftroxgpx/build/2nonpkgs/rust.stuff/rustlearnage/target/debug/deps -L dependency=/home/xftroxgpx/build/2nonpkgs/rust.stuff/rustlearnage/target/debug/deps`
!! LD_LIBRARY_PATH=/home/xftroxgpx/build/2nonpkgs/rust.stuff/rustlearnage/target/debug/deps:/home/xftroxgpx/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib:/home/xftroxgpx/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib:/home/xftroxgpx/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib:
!! executing 'rust/build//x86_64-unknown-linux-gnu/stage2/bin//rustc' with args: '--crate-name bug43197 src/main.rs --crate-type bin --emit=dep-info,link -C debuginfo=2 -C metadata=fd80fe04a98d7b0a -C extra-filename=-fd80fe04a98d7b0a --out-dir /home/xftroxgpx/build/2nonpkgs/rust.stuff/rustlearnage/target/debug/deps -L dependency=/home/xftroxgpx/build/2nonpkgs/rust.stuff/rustlearnage/target/debug/deps'
warning: constant evaluation error: attempt to subtract with overflow. This will become a HARD ERROR in the future
 --> src/main.rs:3:20
  |
3 |     const X: u32 = 33-44;//using src: https://github.com/rust-lang/rust/issues/43197#issuecomment-315016896
  |                    ^^^^^
  |
  = note: #[warn(const_err)] on by default

error[E0080]: constant evaluation error
 --> src/main.rs:3:20
  |
3 |     const X: u32 = 33-44;//using src: https://github.com/rust-lang/rust/issues/43197#issuecomment-315016896
  |                    ^^^^^ attempt to subtract with overflow

error: internal compiler error: src/librustc_trans/mir/constant.rs:396: _1 not initialized
 --> src/main.rs:4:20
  |
4 |     println!("{}", X);//bug
  |                    ^

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.21.0-dev (53bf7903f 2017-07-29) running on x86_64-unknown-linux-gnu

note: run with `RUST_BACKTRACE=1` for a backtrace

!!!!! lvalue=_1 bk=Shared span=src/main.rs:4:20: 4:21
thread 'rustc' panicked at 'Box<Any>', src/librustc_errors/lib.rs:438:8
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
   0: std::sys::imp::backtrace::tracing::imp::unwind_backtrace
             at src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: std::sys_common::backtrace::print
             at src/libstd/sys_common/backtrace.rs:71
             at src/libstd/sys_common/backtrace.rs:60
   2: std::panicking::default_hook::{{closure}}
             at src/libstd/panicking.rs:380
   3: std::panicking::default_hook
             at src/libstd/panicking.rs:390
   4: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:610
   5: std::panicking::begin_panic
             at rust/src/libstd/panicking.rs:571
   6: rustc_errors::Handler::span_bug
             at rust/<panic macros>:3
   7: <std::thread::local::LocalKey<T>>::with
             at rust/src/librustc/session/mod.rs:858
             at rust/src/librustc/ty/context.rs:968
             at rust/src/librustc/ty/context.rs:957
             at rust/src/libstd/thread/local.rs:362
             at rust/src/libstd/thread/local.rs:276
   8: rustc::ty::context::tls::with_opt
             at rust/src/librustc/ty/context.rs:953
             at rust/src/librustc/ty/context.rs:968
   9: rustc::session::span_bug_fmt
             at rust/src/librustc/session/mod.rs:855
             at rust/src/librustc/session/mod.rs:848
  10: rustc_trans::mir::constant::MirConstContext::const_lvalue
             at src/librustc_trans/mir/constant.rs:396
             at rust/src/libcore/option.rs:370
             at src/librustc_trans/mir/constant.rs:395
  11: rustc_trans::mir::constant::MirConstContext::const_rvalue
             at src/librustc_trans/mir/constant.rs:729
  12: rustc_trans::mir::constant::MirConstContext::trans
             at src/librustc_trans/mir/constant.rs:281
  13: rustc_trans::mir::constant::<impl rustc_trans::mir::MirContext<'a, 'tcx>>::trans_constant
             at src/librustc_trans/mir/constant.rs:954
  14: rustc_trans::mir::operand::<impl rustc_trans::mir::MirContext<'a, 'tcx>>::trans_operand
             at src/librustc_trans/mir/operand.rs:298
  15: rustc_trans::mir::rvalue::<impl rustc_trans::mir::MirContext<'a, 'tcx>>::trans_rvalue_operand
             at src/librustc_trans/mir/rvalue.rs:472
  16: rustc_trans::mir::trans_mir
             at src/librustc_trans/mir/statement.rs:38
             at src/librustc_trans/mir/block.rs:49
             at src/librustc_trans/mir/mod.rs:324
  17: rustc_trans::base::trans_instance
             at src/librustc_trans/base.rs:608
  18: rustc_trans::trans_item::TransItem::define
             at src/librustc_trans/trans_item.rs:90
  19: rustc_trans::base::trans_crate::module_translation
             at src/librustc_trans/base.rs:1064
  20: rustc::dep_graph::graph::DepGraph::with_task
             at rust/src/librustc/dep_graph/graph.rs:125
  21: rustc_trans::base::trans_crate
             at src/librustc_trans/base.rs:994
             at rust/src/libcore/ops/function.rs:191
             at rust/src/libcore/option.rs:398
             at rust/src/libcore/iter/mod.rs:1073
             at rust/src/liballoc/vec.rs:1825
             at rust/src/liballoc/vec.rs:1808
             at rust/src/liballoc/vec.rs:1695
             at rust/src/libcore/iter/iterator.rs:1302
             at src/librustc_trans/base.rs:989
  22: rustc_driver::driver::phase_4_translate_to_llvm
             at src/librustc_driver/driver.rs:1070
             at rust/src/librustc/util/common.rs:48
             at src/librustc_driver/driver.rs:1068
  23: rustc_driver::driver::compile_input::{{closure}}
             at src/librustc_driver/driver.rs:211
  24: <std::thread::local::LocalKey<T>>::with
             at src/librustc_driver/driver.rs:1050
             at rust/src/librustc/ty/context.rs:941
             at rust/src/libstd/thread/local.rs:362
             at rust/src/libstd/thread/local.rs:276
  25: rustc::ty::context::tls::enter
             at rust/src/librustc/ty/context.rs:938
  26: <std::thread::local::LocalKey<T>>::with
             at rust/src/librustc/ty/context.rs:925
             at rust/src/libstd/thread/local.rs:362
             at rust/src/libstd/thread/local.rs:276
  27: rustc::ty::context::TyCtxt::create_and_enter
             at rust/src/librustc/ty/context.rs:922
             at rust/src/librustc/ty/context.rs:703
  28: rustc_driver::driver::compile_input
             at src/librustc_driver/driver.rs:963
             at src/librustc_driver/driver.rs:177
  29: rustc_driver::run_compiler
             at src/librustc_driver/lib.rs:218

error: Could not compile `bug43197`.
```

Note: I trimmed absolute paths to avoid some github nowrap in stacktrace lines.
</details>

--- a/src/librustc_errors/diagnostic_builder.rs
+++ b/src/librustc_errors/diagnostic_builder.rs
@@ -82,26 +82,27 @@ impl<'a> DiagnosticBuilder<'a> {
             return;
         }
 
-        match self.level {
+        let is_error = match self.level {
             Level::Bug |
             Level::Fatal |
             Level::PhaseFatal |
             Level::Error => {
-                self.handler.bump_err_count();
+                true
             }
 
             Level::Warning |
             Level::Note |
             Level::Help |
             Level::Cancelled => {
+                false
             }
-        }
+        };
 
         self.handler.emitter.borrow_mut().emit(&self);
         self.cancel();
 
-        if self.level == Level::Error {
-            self.handler.panic_if_treat_err_as_bug();
+        if is_error {
+            self.handler.bump_err_count();
         }
 
         // if self.is_fatal() {
@@ -210,4 +211,3 @@ impl<'a> Drop for DiagnosticBuilder<'a> {
         }
     }
 }
-

--- a/src/librustc_errors/lib.rs
+++ b/src/librustc_errors/lib.rs
@@ -399,7 +399,6 @@ impl Handler {
 
     pub fn span_fatal<S: Into<MultiSpan>>(&self, sp: S, msg: &str) -> FatalError {
         self.emit(&sp.into(), msg, Fatal);
-        self.panic_if_treat_err_as_bug();
         FatalError
     }
     pub fn span_fatal_with_code<S: Into<MultiSpan>>(&self,
@@ -408,12 +407,10 @@ impl Handler {
                                                     code: &str)
                                                     -> FatalError {
         self.emit_with_code(&sp.into(), msg, code, Fatal);
-        self.panic_if_treat_err_as_bug();
         FatalError
     }
     pub fn span_err<S: Into<MultiSpan>>(&self, sp: S, msg: &str) {
         self.emit(&sp.into(), msg, Error);
-        self.panic_if_treat_err_as_bug();
     }
     pub fn mut_span_err<'a, S: Into<MultiSpan>>(&'a self,
                                                 sp: S,
@@ -425,7 +422,6 @@ impl Handler {
     }
     pub fn span_err_with_code<S: Into<MultiSpan>>(&self, sp: S, msg: &str, code: &str) {
         self.emit_with_code(&sp.into(), msg, code, Error);
-        self.panic_if_treat_err_as_bug();
     }
     pub fn span_warn<S: Into<MultiSpan>>(&self, sp: S, msg: &str) {
         self.emit(&sp.into(), msg, Warning);
@@ -494,6 +490,7 @@ impl Handler {
     }
 
     pub fn bump_err_count(&self) {
+        self.panic_if_treat_err_as_bug();
         self.err_count.set(self.err_count.get() + 1);
     }
 

--- a/src/librustc_trans/mir/constant.rs
+++ b/src/librustc_trans/mir/constant.rs
@@ -222,15 +222,24 @@ struct MirConstContext<'a, 'tcx: 'a> {
     substs: &'tcx Substs<'tcx>,
 
     /// Values of locals in a constant or const fn.
-    locals: IndexVec<mir::Local, Option<Const<'tcx>>>
+    locals: IndexVec<mir::Local, Option<Result<Const<'tcx>, ConstEvalErr<'tcx>>>>
 }
 
+fn add_err<'tcx, U, V>(failure: &mut Result<U, ConstEvalErr<'tcx>>,
+                       value: &Result<V, ConstEvalErr<'tcx>>)
+{
+    if let &Err(ref err) = value {
+        if failure.is_ok() {
+            *failure = Err(err.clone());
+        }
+    }
+}
 
 impl<'a, 'tcx> MirConstContext<'a, 'tcx> {
     fn new(ccx: &'a CrateContext<'a, 'tcx>,
            mir: &'a mir::Mir<'tcx>,
            substs: &'tcx Substs<'tcx>,
-           args: IndexVec<mir::Local, Const<'tcx>>)
+           args: IndexVec<mir::Local, Result<Const<'tcx>, ConstEvalErr<'tcx>>>)
            -> MirConstContext<'a, 'tcx> {
         let mut context = MirConstContext {
             ccx: ccx,
@@ -249,7 +258,7 @@ impl<'a, 'tcx> MirConstContext<'a, 'tcx> {
     fn trans_def(ccx: &'a CrateContext<'a, 'tcx>,
                  def_id: DefId,
                  substs: &'tcx Substs<'tcx>,
-                 args: IndexVec<mir::Local, Const<'tcx>>)
+                 args: IndexVec<mir::Local, Result<Const<'tcx>, ConstEvalErr<'tcx>>>)
                  -> Result<Const<'tcx>, ConstEvalErr<'tcx>> {
         let instance = monomorphize::resolve(ccx.shared(), def_id, substs);
         let mir = ccx.tcx().instance_mir(instance.def);
@@ -278,10 +287,9 @@ impl<'a, 'tcx> MirConstContext<'a, 'tcx> {
                     mir::StatementKind::Assign(ref dest, ref rvalue) => {
                         let ty = dest.ty(self.mir, tcx);
                         let ty = self.monomorphize(&ty).to_ty(tcx);
-                        match self.const_rvalue(rvalue, ty, span) {
-                            Ok(value) => self.store(dest, value, span),
-                            Err(err) => if failure.is_ok() { failure = Err(err); }
-                        }
+                        let value = self.const_rvalue(rvalue, ty, span);
+                        add_err(&mut failure, &value);
+                        self.store(dest, value, span);
                     }
                     mir::StatementKind::StorageLive(_) |
                     mir::StatementKind::StorageDead(_) |
@@ -301,9 +309,9 @@ impl<'a, 'tcx> MirConstContext<'a, 'tcx> {
                 mir::TerminatorKind::Goto { target } => target,
                 mir::TerminatorKind::Return => {
                     failure?;
-                    return Ok(self.locals[mir::RETURN_POINTER].unwrap_or_else(|| {
+                    return self.locals[mir::RETURN_POINTER].clone().unwrap_or_else(|| {
                         span_bug!(span, "no returned value in constant");
-                    }));
+                    });
                 }
 
                 mir::TerminatorKind::Assert { ref cond, expected, ref msg, target, .. } => {
@@ -342,33 +350,30 @@ impl<'a, 'tcx> MirConstContext<'a, 'tcx> {
 
                     let mut arg_vals = IndexVec::with_capacity(args.len());
                     for arg in args {
-                        match self.const_operand(arg, span) {
-                            Ok(arg) => { arg_vals.push(arg); },
-                            Err(err) => if failure.is_ok() { failure = Err(err); }
-                        }
+                        let arg_val = self.const_operand(arg, span);
+                        add_err(&mut failure, &arg_val);
+                        arg_vals.push(arg_val);
                     }
                     if let Some((ref dest, target)) = *destination {
-                        if fn_ty.fn_sig(tcx).abi() == Abi::RustIntrinsic {
-                            let value = match &tcx.item_name(def_id).as_str()[..] {
+                        let result = if fn_ty.fn_sig(tcx).abi() == Abi::RustIntrinsic {
+                            match &tcx.item_name(def_id).as_str()[..] {
                                 "size_of" => {
                                     let llval = C_uint(self.ccx,
                                         self.ccx.size_of(substs.type_at(0)));
-                                    Const::new(llval, tcx.types.usize)
+                                    Ok(Const::new(llval, tcx.types.usize))
                                 }
                                 "min_align_of" => {
                                     let llval = C_uint(self.ccx,
                                         self.ccx.align_of(substs.type_at(0)));
-                                    Const::new(llval, tcx.types.usize)
+                                    Ok(Const::new(llval, tcx.types.usize))
                                 }
                                 _ => span_bug!(span, "{:?} in constant", terminator.kind)
-                            };
-                            self.store(dest, value, span);
-                        } else {
-                            match MirConstContext::trans_def(self.ccx, def_id, substs, arg_vals) {
-                                Ok(value) => self.store(dest, value, span),
-                                Err(err) => if failure.is_ok() { failure = Err(err); }
                             }
-                        }
+                        } else {
+                            MirConstContext::trans_def(self.ccx, def_id, substs, arg_vals)
+                        };
+                        add_err(&mut failure, &result);
+                        self.store(dest, result, span);
                         target
                     } else {
                         span_bug!(span, "diverging {:?} in constant", terminator.kind);
@@ -379,7 +384,10 @@ impl<'a, 'tcx> MirConstContext<'a, 'tcx> {
         }
     }
 
-    fn store(&mut self, dest: &mir::Lvalue<'tcx>, value: Const<'tcx>, span: Span) {
+    fn store(&mut self,
+             dest: &mir::Lvalue<'tcx>,
+             value: Result<Const<'tcx>, ConstEvalErr<'tcx>>,
+             span: Span) {
         if let mir::Lvalue::Local(index) = *dest {
             self.locals[index] = Some(value);
         } else {
@@ -392,9 +400,9 @@ impl<'a, 'tcx> MirConstContext<'a, 'tcx> {
         let tcx = self.ccx.tcx();
 
         if let mir::Lvalue::Local(index) = *lvalue {
-            return Ok(self.locals[index].unwrap_or_else(|| {
+            return self.locals[index].clone().unwrap_or_else(|| {
                 span_bug!(span, "{:?} not initialized", lvalue)
-            }).as_lvalue());
+            }).map(|v| v.as_lvalue());
         }
 
         let lvalue = match *lvalue {

--- a/src/test/run-make/treat-err-as-bug/Makefile
+++ b/src/test/run-make/treat-err-as-bug/Makefile
@@ -1,0 +1,5 @@
+-include ../tools.mk
+
+all:
+	$(RUSTC) err.rs -Z treat-err-as-bug 2>&1 \
+	    | grep -q "panicked at 'encountered error with .-Z treat_err_as_bug'"

--- a/src/test/run-make/treat-err-as-bug/err.rs
+++ b/src/test/run-make/treat-err-as-bug/err.rs
@@ -1,0 +1,13 @@
+// Copyright 2017 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+#![crate_type="rlib"]
+
+pub static C: u32 = 0-1;

--- a/src/test/ui/const-eval/issue-43197.rs
+++ b/src/test/ui/const-eval/issue-43197.rs
@@ -1,0 +1,21 @@
+// Copyright 2017 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+#![feature(const_fn)]
+
+const fn foo(x: u32) -> u32 {
+    x
+}
+
+fn main() {
+    const X: u32 = 0-1;
+    const Y: u32 = foo(0-1);
+    println!("{} {}", X, Y);
+}

--- a/src/test/ui/const-eval/issue-43197.stderr
+++ b/src/test/ui/const-eval/issue-43197.stderr
@@ -1,0 +1,28 @@
+warning: constant evaluation error: attempt to subtract with overflow. This will become a HARD ERROR in the future
+  --> $DIR/issue-43197.rs:18:20
+   |
+18 |     const X: u32 = 0-1;
+   |                    ^^^
+   |
+   = note: #[warn(const_err)] on by default
+
+warning: constant evaluation error: attempt to subtract with overflow. This will become a HARD ERROR in the future
+  --> $DIR/issue-43197.rs:19:20
+   |
+19 |     const Y: u32 = foo(0-1);
+   |                    ^^^^^^^^
+
+error[E0080]: constant evaluation error
+  --> $DIR/issue-43197.rs:18:20
+   |
+18 |     const X: u32 = 0-1;
+   |                    ^^^ attempt to subtract with overflow
+
+error[E0080]: constant evaluation error
+  --> $DIR/issue-43197.rs:19:24
+   |
+19 |     const Y: u32 = foo(0-1);
+   |                        ^^^ attempt to subtract with overflow
+
+error: aborting due to 2 previous errors
+

